network/transport/tcp_socket.c

```diff
@@ -45,23 +45,169 @@
 /*************************** Types Declarations *******************************/
 /******************************************************************************/
 
+/**
+ * @struct secure_socket_desc
+ * @brief Fields used by secure socket
+ */
+struct secure_socket_desc {
+	/** True random number generator reference */
+	struct trng_desc		*trng;
+	/* Mbed structures */
+	/** Certification authority certificate */
+	mbedtls_x509_crt		ca_cert;
+	/** Client certificate */
+	mbedtls_x509_crt		client_cert;
+	/** SSL configuration structure */
+	mbedtls_ssl_config		conf;
+	/** Mbedtls tls context */
+	mbedtls_ssl_context		ssl;
+};
+
 /* Socket descriptor */
 struct tcp_socket_desc {
 	/* Id of the opened socket */
 	uint32_t			id;
 	/* Reference to the network interface */
 	struct network_interface	*net;
+	/* Reference to secure descriptor */
+	struct secure_socket_desc	*secure;
 };
 
 /******************************************************************************/
 /************************ Functions Definitions *******************************/
 /******************************************************************************/
 
+/* Wrapper over socket_recv */
+static int tls_net_recv(struct tcp_socket_desc *sock, unsigned char *buff,
+			size_t len)
+{
+	int32_t ret;
+
+	ret = sock->net->socket_recv(sock->net->net, sock->id, buff, len);
+	if (ret == -EAGAIN)
+		return MBEDTLS_ERR_SSL_WANT_READ;
+
+	return ret;
+}
+
+/* Wrapper over socket_recv */
+static int tls_net_send(struct tcp_socket_desc *sock, unsigned char *buff,
+			size_t len)
+{
+	return sock->net->socket_send(sock->net->net, sock->id, buff, len);
+}
+
+/* Remove secure descriptor*/
+static void stcp_socket_remove(struct secure_socket_desc *desc)
+{
+	mbedtls_x509_crt_free(&desc->ca_cert);
+	mbedtls_x509_crt_free(&desc->client_cert);
+	mbedtls_ssl_config_free(&desc->conf);
+	if (desc->trng)
+		trng_remove(desc->trng);
+	free(desc);
+}
+
+/* Init secure descriptor */
+static int32_t stcp_socket_init(struct secure_socket_desc **desc,
+				struct tcp_socket_desc *sock,
+				struct secure_init_param *param)
+{
+	struct secure_socket_desc	*ldesc;
+	int32_t				ret;
+
+	if (!desc || !param || !param->ca_cert)
+		return FAILURE;
+
+	ldesc = (typeof(ldesc))calloc(1, sizeof(*ldesc));
+	if (!ldesc)
+		return FAILURE;
+
+	/* Initialize structures */
+	mbedtls_x509_crt_init(&ldesc->ca_cert);
+	mbedtls_x509_crt_init(&ldesc->client_cert);
+	mbedtls_ssl_config_init(&ldesc->conf);
+
+	ret = trng_init(&ldesc->trng, param->trng_init_param);
+	if (IS_ERR_VALUE(ret)) {
+		ldesc->trng = NULL;
+		goto exit;
+	}
+
+	/* Set default configuration: TLS client socket */
+	ret = mbedtls_ssl_config_defaults(&ldesc->conf,
+					  MBEDTLS_SSL_IS_CLIENT,
+					  MBEDTLS_SSL_TRANSPORT_STREAM,
+					  MBEDTLS_SSL_PRESET_DEFAULT);
+	if (IS_ERR_VALUE(ret))
+		goto exit;
+
+	/* Certificate validation is set as mandatory */
+	mbedtls_ssl_conf_authmode(&ldesc->conf, MBEDTLS_SSL_VERIFY_NONE);
```

Comment:
If certificate check is mandatory and this is a client shouldn't the parameter be `MBEDTLS_SSL_VERIFY_REQUIRED`? [Function comments](https://github.com/ARMmbed/mbedtls/blob/5d318cc22b30235d580528c85b7e52e8c565695c/include/mbedtls/ssl.h#L1482) seem to suggest so.

Reply:
It didn't work with verification (maybe I have to fix some configuration).
It is in the next steps from https://github.com/analogdevicesinc/no-OS/pull/586
I will solve this and the rest in other pull request.
This is a first functional version.

Reply:
Maybe you need to call `mbedtls_ssl_config_init` after `mbedtls_ssl_config_defaults` to actually set the config.

Reply:
init should be called befora any use of this structures. It just set the memory to 0. This is how is done in the library examples also.